rfcs/009 - Tree RFC.md

```diff
@@ -0,0 +1,110 @@
+
+An equivalent to `map` for Tree makes sense, letting you transform the values of one tree into a new one, where the callback function could also use the parent and/or children to determine how to create the newly transformed node. The only question is whether or not it uses the same name of `map` or it gets a new name to be clearer?
+
+`reduce` is similarly interesting -- it could be parallelized by "dependency layer", all of the nodes with no children, then all of the nodes in the first layer above those, then the layer above that, etc, until you reach the root node. This means that serializing a data structure built on top of a Tree (like JSON, for instance) would be an automatically parallel operation and could therefore be more performant by default than other languages.
+
+`filter` could have two implementations, one that filters in a way that a failed filter node *and all of its children* are removed, and one where the failed nodes are deleted and their children (that pass the filter) are re-parented on the failed node's parent, where the latter would provide an array of potentially more than one tree if the root node is filtered away. What they should be called is not clear.
+
+`every` and `some` could simply be shorthand for something like `someTree.toNodeArray().every(...)` or just excluded instead.
+
+There are also operations that are purely tree-related operations like a `balance` to rebalance a tree as a binary tree (though this would be part of also creating an array `sort` function and defining an `Orderable` interface type).
+
+It should also be possible to provide syntactic sugar for statically-defined Trees, inspired by XML:
+
+```ln
+const myTree = <"foo">
+  <"bar">
+    <"baz"/>
+  </>
+  <"bay"/>
+</> endTree
+```
+
+where the output type in this example is `Tree<string>`. It requires 4 operators, `<`, `>`, `/>`, and `</>`. A prefix `<` causes it to create a new tree with a root node with the value that follows it and has the highest precedence. `>` is an infix operator between two trees or a node and a tree and attaches the second argument as a child of the first argument (a node or the root node of the tree). `/>` and `</>` are prefix operators that create a special intermediate type that causes traversal backwards up the first node before attaching, to get the correct nesting set up except the final one as an infix operator between the two nodes to actually join. That final `endTree` is simply an alias to `void` that tells it to finalize the Tree structure and simply return it and is necessary because Alan does not support postfix operators.
+
+This is the iffiest idea because it's done through an abuse of the operator syntax, but it looks pretty natural for defining tree constants. Fluent style isn't so bad, though:
+
+```ln
+const myTree = newTree("foo")
+  .addChild(newTree("bar")
+    .addChild("baz"))
+  .addChild("bay")
+```
+
+With the only downside being keeping track of the parens.
```

Comment:
This should go first since this will be the default way of defining a tree initially